libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,717 @@
+        response_payload->size = 0;
+        return kErrNone;
+    }
+    else {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_CreateDirectory(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    if (mkdir(_ftp_current_abs_path, S_IRWXU | S_IRWXG | S_IRWXO) == 0) {
+        response_payload->size = 0;
+        return kErrNone;
+    }
+    else {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_RemoveFile(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    if (unlink(_ftp_current_abs_path) == 0) {
+        response_payload->size = 0;
+        return kErrNone;
+    }
+    else {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_TruncateFile(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    int ret = truncate(_ftp_current_abs_path, request_payload->offset);
+
+    if (ret == 0) {
+        return kErrNone;
+    }
+    _ftp_errno = errno;
+    return kErrFailErrno;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Rename(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    char *ptr = _data_as_cstring(request_payload);
+    size_t oldpath_sz = strlen(ptr);
+
+    if (oldpath_sz >= request_payload->size) {
+        // no newpath
+        _ftp_errno = EINVAL;
+        return kErrFailErrno;
+    }
+
+    _set_current_abs_path(ptr);
+    char* current_path = new char[strlen(_ftp_current_abs_path)];
+    strcpy(current_path, _ftp_current_abs_path);
+
+    _set_current_abs_path(ptr + oldpath_sz + 1);
+    _ftp_current_abs_path[sizeof(_ftp_current_abs_path) - 1] = '\0'; // ensure termination
+
+    if (rename(current_path, _ftp_current_abs_path) == 0) {
+        response_payload->size = 0;
+        return kErrNone;
+
+    }
+    _ftp_errno = errno;
+    return kErrFailErrno;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_CalcFileCRC32(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    uint32_t checksum = 0;
+    ssize_t bytes_read;
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    int fd = ::open(_ftp_current_abs_path, O_RDONLY);
+
+    if (fd < 0) {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+
+    do {
+        bytes_read = ::read(fd, response_payload->data, MAX_FTP_DATA_LENGTH);
+
+        if (bytes_read < 0) {
+            _ftp_errno = errno;
+            ::close(fd);
+            return kErrFailErrno;
+        }
+        checksum = crc_crc32(checksum, (uint8_t *)response_payload->data, bytes_read);
+    }
+    while (bytes_read == MAX_FTP_DATA_LENGTH);
```

Comment:
this could take a very very long time with a large file. If someone does a ftp op while flying I think it's likely their vehicle will crash due to this CRC function. We can't do this in the main thread